codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java

```diff
@@ -214,15 +234,80 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
             dstPort = header.readUnsignedShort();
         }
 
-        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort);
+        final List<HAProxyTLV> tlvs = readTlvs(header);
+
+        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort, tlvs);
+    }
+
+    private static List<HAProxyTLV> readTlvs(final ByteBuf header) {
+        HAProxyTLV haProxyTLV = readNextTLV(header);
+        if (haProxyTLV == null) {
+            return Collections.emptyList();
+        }
+        // In most cases there are less than 4 TLVs available
+        List<HAProxyTLV> haProxyTLVs = new ArrayList<HAProxyTLV>(4);
+
+        do {
+            haProxyTLVs.add(haProxyTLV.retain());
+            if (haProxyTLV instanceof HAProxySSLTLV) {
+                haProxyTLVs.addAll(((HAProxySSLTLV) haProxyTLV).encapsulatedTLVs());
+            }
+        } while ((haProxyTLV = readNextTLV(header)) != null);
+        return haProxyTLVs;
+    }
+
+    private static HAProxyTLV readNextTLV(final ByteBuf header) {
+
+        // We need at least 4 bytes for a TLV
+        if (header.readableBytes() < 4) {
+            return null;
+        }
+
+        final byte typeAsByte = header.readByte();
+        final HAProxyTLV.Type type = HAProxyTLV.Type.typeForByteValue(typeAsByte);
+
+        final int length = header.readUnsignedShort();
+        switch (type) {
```

Comment:
hmm the intelliJ code inspection said to remove the default branch
![bildschirmfoto 2017-01-19 um 09 34 05](https://cloud.githubusercontent.com/assets/10121608/22099184/8a6b7f80-de2a-11e6-8987-543d43138a88.png)